.github/workflows/ci.yml

```diff
@@ -45,26 +50,67 @@ jobs:
         with:
           fetch-depth: 0
       - name: Restore npm dependencies
-        uses: actions/cache@v4
-        id: npm-cache
+        uses: actions/cache/restore@v4
         with:
           path: |
             node_modules
-          key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
+          key: ${{ runner.os }}-npm-${{matrix.node}}-${{ hashFiles('**/package-lock.json') }}
           restore-keys: |
+            ${{ runner.os }}-npm-${{matrix.node}}
             ${{ runner.os }}-npm
-      - run: npm run build
       - run: npx playwright install
+      - run: npm run build
       - name: e2e
-        run: npm run ci:e2e
-      - uses: actions/upload-artifact@v3
+        run: npm run ci:e2e -- --shard=${{ matrix.shardIndex }}/${{ matrix.shardTotal }}
+      - name: Upload blob report to GitHub Actions Artifacts
         if: always()
+        uses: actions/upload-artifact@v4
         with:
-          name: playwright-report
-          path: playwright-report/
-          retention-days: 30
+          name: blob-report-${{runner.os}}-${{matrix.node}}-${{ matrix.shardIndex }}
```

Comment:
this was annoying but this is how we can make unique caches across all these machines / node versions